### ISSUE TITLE
Failed - Updating/Deploying Azure custom roles definition through terraform

### ISSUE THREAD

**Issue by sk-devsecops** [hashicorp/terraform-provider-azurerm#14094]
We are using terraform to deploy/update Azure custom roles. we are using azure devops CICD pipeline and Azure CLI task to run terraform apply. Terraform got intermittent issue, when we update a custom role, the terraform apply command run and try to update the Azure custom role but terraform keeps waiting to get response back from Azure and show following message 
**Still modifying...** - please refer image below
Task run for 60 minutes and terraform keeps showing "Still modifying..." message and task time out after 60 mins and pipeline failed. It happens after every second or third time. Sometime it works fine and update Azure roles within 10 - 11mins. 

Note - When we check Azure custom roles from Azure portal, the roles gets updated but terraform apply task never completed and shows "Still modifying..." message in CICD pipeline. Seems like terraform task never gets respond back from Azure even after updating Azure roles successfully, hence task never ends.

### Terraform (and AzureRM Provider) Version

Terraform v1.0.10
hashicorp/azurerm v2.84.0
hashicorp/azuread v2.8.0

### Affected Resource(s)

Azure custom roles definition.
Azure role assignment

* `azurerm_v2.84.0`

### Terraform Configuration Files

terraform {
  required_providers {
    azuread = {
      source  = "hashicorp/azuread"
      version = "~> 2.8.0"
    }
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 2.84.0"
    }
  }
  backend "azurerm" {
  }
}

provider "azurerm" {
  features {}
}

provider "azuread" {
}

### Debug Output
![image](https://user-images.githubusercontent.com/93953979/140837607-ae771008-8799-4dcc-a3d9-86b59c847840.png)
![image](https://user-images.githubusercontent.com/93953979/140838055-692df4e6-f7d7-4d75-bdd8-02895839f5dc.png)

### Expected Behaviour

Terraform apply task should complete successfully. Following is the expected result. Sometime it works file and task complete successfully
![image](https://user-images.githubusercontent.com/93953979/140839077-18d07e80-a65c-4aa2-8fc5-94dbcb3c7cf8.png)

### Actual Behaviour

Terraform apply task did not complete within time frame. Seems like azure does not respond back to terraform apply. Please check above image

### Steps to Reproduce

Use Azure CICD pipeline to run terraform
Add/Remove actions from the Azure custom role
terraform init
terraform apply

1. `terraform apply`

### Important Factoids

Azure resources running in Australia East and Australia South East

### References

* #0000

**Comment by aristosvo** [comment]
Hi @sk-devsecops! I ran into the same issue when debugging another issue.

It seems there is in some cases (and I'm trying to find out which cases) no change induced in Azure, but a change is seen in Terraform. When Terraform tries to update Azure, the change is ignored by Azure, but (and this is specific for this resource) we check whether the resource is actually updated by checking the timestamp.

As Azure does not change, this timestamp check will loop until timeout..

My setup which shows this error after second apply:

```hcl
provider "azurerm" {
    features {}
}

data "azurerm_subscription" "primary" {
}

resource "azurerm_role_definition" "test_role" {
  name  = "test_role_1"
  scope = data.azurerm_subscription.primary.id
  permissions {
    actions     = []
    not_actions = []
  }
  assignable_scopes = [
    data.azurerm_subscription.primary.id,
  ]
}
data "azurerm_role_definition" "select_test_role" {
  name  = "test_role_1"
  scope = data.azurerm_subscription.primary.id
  depends_on = [
    azurerm_role_definition.test_role,
  ]
}
```

**Comment by sk-devsecops** [comment]
Hi @aristosvo, I do not have any empty blocks. It works fine occasionally but most of the time it failed. I reproduce the issue on my local machine without using any storage account for state file. I am using a local state file and added logging. I got following error in the log file.

![image](https://user-images.githubusercontent.com/93953979/141242637-465a3a5a-8425-4d89-9514-e4ceababbf63.png)

2021-11-11T13:15:36.126+1100 [TRACE] vertex "azurerm_role_definition.soc_analyst_l1": visit complete
**2021-11-11T13:15:36.126+1100 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" errored, so skipping**
2021-11-11T13:15:36.126+1100 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2021-11-11T13:15:36.127+1100 [TRACE] dag/walk: upstream of "root" errored, so skipping
2021-11-11T13:15:36.127+1100 [TRACE] statemgr.Filesystem: have already backed up original ./sentinelroles.test.tfstate to ./sentinelroles.test.tfstate.backup on a previous write
2021-11-11T13:15:36.142+1100 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 53
2021-11-11T13:15:36.142+1100 [TRACE] statemgr.Filesystem: writing snapshot at ./sentinelroles.test.tfstate
2021-11-11T13:15:36.159+1100 [TRACE] statemgr.Filesystem: removing lock metadata file .sentinelroles.test.tfstate.lock.info
2021-11-11T13:15:36.162+1100 [TRACE] statemgr.Filesystem: unlocking ./sentinelroles.test.tfstate using fcntl flock
**2021-11-11T13:15:36.168+1100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"**
2021-11-11T13:15:36.179+1100 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/azurerm/2.84.0/linux_amd64/terraform-provider-azurerm_v2.84.0_x5 pid=7187
2021-11-11T13:15:36.179+1100 [DEBUG] provider: plugin exited

**Comment by aristosvo** [comment]
Can you show the `plan` for an occasion that it times out?

**Comment by LaurentLesle** [comment]
Yes facing the same issue. Timeout after 60 mins

![image](https://user-images.githubusercontent.com/4702224/142569634-49eea9d1-8d33-4277-be79-bc7c618a8a99.png)

**Comment by sk-devsecops** [comment]
My plan show that new action will be added to 4 role definitions as shown in image, thats it
![image](https://user-images.githubusercontent.com/93953979/145343735-2c5c616e-c99a-492b-8fdc-e0a9a0ed6c0c.png)

I try using the latest version as well. Still the same issue
hashicorp/azurerm v2.88.1
hashicorp/azuread v2.12.0

**Comment by mikebranstein** [comment]
I too have this issue. It's intermittent, as @sk-devsecops has mentioned. 

In my code base where I see this issue, we create custom roles and allow them to be assigned at the subscription scope. We frequently add newly-created subscriptions to the assignable_scopes list. I have ~10 custom roles, and the first 8 or 9 cusotm roles seem to update without a timeout. The last 1-2 roles timeout my pipeline after 30 minutes. When I inspect the custom roles in Azure, they have been updated. 

I've noticed this issue on and off for the last year, but updating to 2.77.0 seemed to resolve the issue. Until recently, it's begun happening again.

**Comment by mikebranstein** [comment]
@sk-devsecops and @aristosvo - I've been digging into the provider code and believe I've found the reason why we're seeing this issue. 

### Summary

I believe there is a logic error or edge case that is not being accounted for correctly in the role definition resource code. I believe there is either an invalid assumption in the "update date" value of a role definition, or Azure updates a role definition so quickly that the provider code never detects the change in the "update date", thus creating an infinite wait loop. The specific lines of code are: https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L391-L394

My environment setup is similar to @sk-devsecops: I create and update custom role definitions. My setup differs in a few ways though: 
- I do not typically add or remove permissions to the customer roles. 
- I regularly add additional subscriptions to the assignable_scopes property, as I create more subscriptions in the tenant.
 
I have seen the issue occur with custom roles that have empty properties defined (i.e., not_action = []), and with custom roles that do not include empty permissions (i.e., not_action = [] is not present in the permissions block). 

In summary, I believe this isn't related to these properties as much as it's related to how the provider checks for an updated resource. See my supporting research below.

### Background

When role definitions are updated, the _resourceArmRoleDefinitionUpdate_ function is called:

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L214

In the comments of the function, the author calls out a peculiar implementation details of the AzureRM REST API for updating role definitions:

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L255-L258

I read this comments as Azure doesn't actually "update" role definitions - it creates a new role definition, then swaps the old for the new. During this "swap" process, the "create date" and "update date" properties display an odd behavior. First, the "create date" and "update date" display the same timestamp (which was when Azure created the new). Second, the old and new resources coalesce, and the "create date" of the new resource is updated to the "create date" value of the old resource.

Prior to this comment the provider makes the update call to the AzureRM REST API, then stores off the "update date" returned from the API. The provider then passes this date to the function responsible for long-polling the API to detect when the resource has been officially "updated":

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L250-L268

You'll notice the function _roleDefinitionEventualConsistencyUpdate_ is used to determine when the resource has updated:

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L266

When you drill into this function, the first ~30 lines perform error checking and parsing. The last 10 lines determine whether the resource is still updating "Pending" or if it has been updated "Updated":

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L386-L396

Lines 398-389 address the first period of the swap, where the create date and update date are equal:

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L386-L389

Lines 391-394 address another case where the resource's current update date is compared to the update date captured from the original REST API call:

https://github.com/hashicorp/terraform-provider-azurerm/blob/763f515974500e358cd64ed2a995adfa8c73ba96/internal/services/authorization/role_definition_resource.go#L391-L394

This second conditional can be misleading or problematic in a few ways:
- The variable _updateRequestTime_ shouldn't be read as "time of update request". It is actually the value of the "update date" as returned by the original REST API call. _updateRequestTime_ may be more meaningful if it were the value of the resource "update date" BEFORE any updates are made to it.
- There's an assumption in the code that the update date will be changed from the time the Azure API responds to the original update request and when the provider checks back in over the next several minutes. This may not be guaranteed - what if the update operation happens extremely fast, and when the REST API returns, the resource has already been updated and the "update date" will no longer change in value.

In conclusion, I believe this section of code is responsible for the behavior we're seeing.

**Comment by rcskosir** [comment]
Thanks for opening this issue. This was a problem in the 2.x version of the provider which is no longer actively maintained. This was likely fixed by PR #15817.
If this is still an issue with the 3.x version of the provider please do let us know by opening a new issue, thanks!
